### ISSUE TITLE
Embed / Optional issue

### ISSUE THREAD

**Issue by lodagro** [construct/construct#28]
In the example below the BitStruct *i* can turn undesired into an embedded one, depending if the two fields defined before it are Embedded and/or Optional. At the bottom of the code there are three tests describing the issue.

```python
import construct


def vstring(name, embed=True, optional=True):
    name_length_field = "_%s_length" % name.lower()
    s = construct.Struct(
        name,
        construct.ULInt8(name_length_field),
        construct.String(name,
                         lambda ctx: getattr(ctx, name_length_field)))
    if optional:
        s = construct.Optional(s)
    if embed:
        s = construct.Embed(s)
    return s


def build_struct(embed_g=True, embed_h=True):
    s = construct.Struct(
        "mystruct",
        construct.ULInt32("a"),
        construct.ULInt8("b"),
        construct.ULInt8("c"),
        construct.BitStruct(
            "d",
            construct.BitField("d_bit7", 1),
            construct.BitField("d_bit6", 1),
            construct.BitField("d_bit5", 1),
            construct.BitField("d_bit4", 1),
            construct.BitField("d_bit3", 1),
            construct.BitField("d_bit2", 1),
            construct.BitField("d_bit1", 1),
            construct.BitField("d_bit0", 1)),
        construct.BitStruct(
            "e",
            construct.BitField("e_bit7", 1),
            construct.BitField("e_bit6", 1),
            construct.BitField("e_bit5", 1),
            construct.BitField("e_bit4", 1),
            construct.BitField("e_bit3", 1),
            construct.BitField("e_bit2", 1),
            construct.BitField("e_bit1", 1),
            construct.BitField("e_bit0", 1)),
        construct.BFloat32("f"),
        vstring("g", embed=embed_g),
        vstring("h", embed=embed_h),
        construct.BitStruct(
            "i",
            construct.BitField("i_bit7", 1),
            construct.BitField("i_bit6", 1),
            construct.BitField("i_bit5", 1),
            construct.BitField("i_bit4", 1),
            construct.BitField("i_bit3", 1),
            construct.BitField("i_bit2", 1),
            construct.BitField("i_bit1", 1),
            construct.BitField("i_bit0", 1)),
        construct.SBInt8("j"),
        construct.SBInt8("k"),
        construct.SBInt8("l"),
        construct.LFloat32("m"),
        construct.LFloat32("n"),
        vstring("o"),
        vstring("p"),
        vstring("q"),
        vstring("r"))
    return s


bytes = ('\xc3\xc0{\x00\x01\x00\x00\x00HOqA\x12some silly text...\x00\x0e\x00'
         '\x00\x00q=jAq=zA\x02dB\x02%f\x02%f\x02%f')

print ("\n\nNo embedding for neither g and h, i is a container --> OK")
print build_struct(embed_g=False, embed_h=False).parse(bytes)
print ("Embed both g and h, i is not a container --> FAIL")
print build_struct(embed_g=True, embed_h=True).parse(bytes)
print ("\n\nEmbed g but not h --> EXCEPTION")
print build_struct(embed_g=True, embed_h=False).parse(bytes)

# When setting optional to False in vstring method, all three tests above
# work fine.
```

running the scripts (python 2.7, construct 2.5.1)

```python
No embedding for neither g and h, i is a container --> OK
Container:
    a = 8110275
    b = 1
    c = 0
    d = Container:
        d_bit7 = 0
        d_bit6 = 0
        d_bit5 = 0
        d_bit4 = 0
        d_bit3 = 0
        d_bit2 = 0
        d_bit1 = 0
        d_bit0 = 0
    e = Container:
        e_bit7 = 0
        e_bit6 = 0
        e_bit5 = 0
        e_bit4 = 0
        e_bit3 = 0
        e_bit2 = 0
        e_bit1 = 0
        e_bit0 = 0
    f = 212421.015625
    g = Container:
        g = 'some silly text...'
    h = Container:
        h = ''
    i = Container:
        i_bit7 = 0
        i_bit6 = 0
        i_bit5 = 0
        i_bit4 = 0
        i_bit3 = 1
        i_bit2 = 1
        i_bit1 = 1
        i_bit0 = 0
    j = 0
    k = 0
    l = 0
    m = 14.640000343322754
    n = 15.640000343322754
    o = 'dB'
    p = '%f'
    q = '%f'
    r = '%f'
Embed both g and h, i is not a container --> FAIL
Container:
    a = 8110275
    b = 1
    c = 0
    d = Container:
        d_bit7 = 0
        d_bit6 = 0
        d_bit5 = 0
        d_bit4 = 0
        d_bit3 = 0
        d_bit2 = 0
        d_bit1 = 0
        d_bit0 = 0
    e = Container:
        e_bit7 = 0
        e_bit6 = 0
        e_bit5 = 0
        e_bit4 = 0
        e_bit3 = 0
        e_bit2 = 0
        e_bit1 = 0
        e_bit0 = 0
    f = 212421.015625
    g = 'some silly text...'
    h = ''
    i_bit7 = 0
    i_bit6 = 0
    i_bit5 = 0
    i_bit4 = 0
    i_bit3 = 1
    i_bit2 = 1
    i_bit1 = 1
    i_bit0 = 0
    i = <...>
    j = 0
    k = 0
    l = 0
    m = 14.640000343322754
    n = 15.640000343322754
    o = 'dB'
    p = '%f'
    q = '%f'
    r = '%f'


Embed g but not h --> EXCEPTION
Traceback (most recent call last):
  File "embed.py", line 77, in <module>
    print build_struct(embed_g=True, embed_h=False).parse(bytes)
  File ".../lib/python2.7/site-packages/construct/core.py", line 187, in parse
    return self.parse_stream(BytesIO(data))
  File ".../lib/python2.7/site-packages/construct/core.py", line 197, in parse_stream
    return self._parse(stream, Container())
  File ".../lib/python2.7/site-packages/construct/core.py", line 664, in _parse
    raise OverwriteError("%r would be overwritten but allow_overwrite is False" % (sc.name,))
construct.core.OverwriteError: 'h' would be overwritten but allow_overwrite is False
```

**Comment by gkonstantyno** [comment]
I also observed this behavior. Fix seems to be easy. When you have mixed embedded and struct constructs when building, original parent <obj> in some circumstances is left and passed to not emmbedded struct. This causes writing same elements twice in struct and in parent. Deletion of overleft parent obj in `_parse` function of `Struct` fixes this.

```python
        for sc in self.subcons:
            if sc.conflags & self.FLAG_EMBED:
                context["<obj>"] = obj
                sc._parse(stream, context)
                if "<obj>" in context:
                    del context["<obj>"]
```

All tests are passing after this change.

**NOTE:** there is also a bug in Container pop function - in case of trying to pop not existing element with default value.

**Comment by arekbulski** [comment]
Could you simpify the test and add it to the test suite through Pull Request? This piece of code is hard to read and testing should be automatic.

**Comment by arekbulski** [comment]
Added a test in new branch, as shown. Now could someone submit a fix?

**Comment by gkonstantyno** [comment]
Please check my pull request

**Comment by arekbulski** [comment]
Fixed in 0a37379db83f2f5bfdad14ddcfd50ae2b3445c51 as submitted by @gkonstantyno . :cake:  Thank you.

**Comment by arekbulski** [comment]
I could use some help with processing issues and features. There are some things that do not work yet and I dont know why. Need assistance.
https://github.com/construct/construct/branches
